zebra-network/src/config.rs

```diff
@@ -71,9 +76,54 @@ pub struct Config {
     /// testnet.
     pub initial_testnet_peers: IndexSet<String>,
 
+    /// An optional root directory for storing cached peer address data.
+    /// If this is `None`, then peer addresses are not read or written to disk.
+    ///
+    /// If you change this directory, you might also want to change `state.cache_dir`.
+    /// By default, all Zebra instances run by the same user will share a single peer cache.
+    ///
+    /// The peer cache is a list of the addresses of some recently useful peers.
+    ///
+    /// For privacy reasons, the cache does *not* include any other information about peers,
+    /// such as when they were connected to the node.
+    ///
+    /// Deleting or modifying the peer cache can impact your node's:
+    /// - reliability: if DNS or the Zcash DNS seeders are unavailable or broken
+    /// - security: if DNS is compromised with malicious peers
+    ///
+    /// If you delete it, Zebra will replace it with a fresh set of peers from the DNS seeders.
+    ///
+    /// The default directory is platform dependent, based on
+    /// [`dirs::cache_dir()`](https://docs.rs/dirs/3.0.1/dirs/fn.cache_dir.html):
+    ///
+    /// |Platform | Value                                           | Example                              |
+    /// | ------- | ----------------------------------------------- | ------------------------------------ |
+    /// | Linux   | `$XDG_CACHE_HOME/zebra` or `$HOME/.cache/zebra` | `/home/alice/.cache/zebra`           |
+    /// | macOS   | `$HOME/Library/Caches/zebra`                    | `/Users/Alice/Library/Caches/zebra`  |
+    /// | Windows | `{FOLDERID_LocalAppData}\zebra`                 | `C:\Users\Alice\AppData\Local\zebra` |
+    /// | Other   | `std::env::current_dir()/cache/zebra`           | `/cache/zebra`                       |
+    ///
+    /// # Security
+    ///
+    /// If you are running Zebra with elevated permissions ("root"), create the
+    /// directory for this file before running Zebra, and make sure the Zebra user
+    /// account has exclusive access to that directory, and other users can't modify
+    /// its parent directories.
+    ///
+    /// # Implementation Details
+    ///
+    /// Each network has a separate peer list, which is updated regularly from the current
+    /// address book. These lists are stored in `network/mainnet.peers` and
+    /// `network/testnet.peers` files, underneath the `cache_dir` path.
+    ///
+    /// Previous peer lists are automatically loaded at startup, and used to populate the
+    /// initial peer set and address book.
+    pub cache_dir: Option<PathBuf>,
```

Comment:
This is wrong because Option::None is not written to disk, and then when it is read back it is replaced by Option::Some(default_path).

Reply:
Use a custom untagged enum with `bool` and `String` (`OsString`?) types.
https://serde.rs/enum-representations.html#untagged

Reply:
This should work:

```rust

#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(untagged)]
pub enum ShouldCachePeers {
    Path(PathBuf),
    Bool(bool),
}

impl ShouldCachePeers {
    pub fn cache_dir(&self) -> Option<PathBuf> {
        match self {
            Self::Path(cache_dir) => Some(cache_dir.to_owned()),
            Self::Bool(should_cache) => should_cache.then(|| {
                dirs::cache_dir()
                    .unwrap_or_else(|| std::env::current_dir().unwrap().join("cache"))
                    .join("zebra")
            }),
        }
    }
}

pub fn peer_cache_file_path(&self, network: Network) -> Option<PathBuf> {
    Some(
        self.should_cache_peers
            .cache_dir()?
            .join("network")
            .join(format!("{}.peers", network.lowercase_name())),
    )
}

```

Reply:
Thanks, I switched some names around, but the overall implementation is the same.